### ISSUE TITLE
PHP suhosin on Debian csv export breaks due to GET var name length limit

### ISSUE THREAD

**Issue by hailwood** [silverstripe/silverstripe-framework#3855]
A default gridfield in a model admin, or the gridfield for a report - clicking the "Export to CSV" or "Print" buttons does not do the specified action - instead you end up with an empty page that contains only the html for the gridfield and a url that looks like 

>[removed]/admin/reports/SideReport_RecentlyEdited/EditForm/field/Report?Report%5BGridState%5D=%7B"GridFieldSortableHeader"%3A%7B"SortColumn"%3A%5B%5D%7D%2C"GridFieldPaginator"%3A%7B"currentPage"%3A1%7D%7D&action_gridFieldAlterAction%3FStateID%3Dd0f05d69129a2cbc8d0cfc4894cdc8fa=Export%20to%20CSV

![screenshot from 2015-02-05 13 18 10](https://cloud.githubusercontent.com/assets/709773/6052596/87138586-ad39-11e4-8027-0512c46b4802.png)

How to reproduce:

- Install silverstripe (~3.1) on server that has php 5.3.3
- Edit a page
- Visit Recently Edited Pages report
- Attempt to export or print

This has been confirmed as not an issue for php 5.3.28 so somewhere between .3 and .28 the issue has been fixed, a workaround is to change framework/forms/gridfield/GridField.php:617 to `parse_str($_SERVER['QUERY_STRING'], $data);`

**Comment by dhensby** [comment]
I have just tested this and it seems to be working for me using Centos 6.5 running php 5.3.3

Steps used to create test environment:

Tested with [vagrant 1.7.2](https://www.vagrantup.com/) and [virtualbox 4.3.20](https://www.virtualbox.org/)

```bash
$ vagrant init centos-65-x64-virtualbox-nocm http://puppet-vagrant-boxes.puppetlabs.com/centos-65-x64-virtualbox-nocm.box
$ sed -i '25i  config.vm.network \"forwarded_port\", guest: 80, host: 8080' Vagrantfile
$ vagrant up
$ vagrant ssh
$ sudo su
$ yum install -y php php-common php-pdo php-mysql php-session php-dom php-gd php-fileinfo php-hash php-iconv php-mbstring php-simplexml php-tokenizer php-xml php-tidy httpd mysql-server git
$ sed -i 's/User apache/User vagrant/i' /etc/httpd/conf/httpd.conf
$ sed -i 's/Group apache/Group vagrant/i' /etc/httpd/conf/httpd.conf
$ sed -i 's/EnableSendfile on/EnableSendfile off/i' /etc/httpd/conf/httpd.conf
$ sed -i '/<Directory \"\/var\/www\/html\">/,/<\/Directory>/ { s/AllowOverride None/AllowOverride All/i }' /etc/httpd/conf/httpd.conf
$ chown -R vagrant: /var/lib/php/session
$ cp -f /usr/share/doc/php-*/php.ini-development /etc/php.ini
$ sed -i 's/;date\.timezone.*/date\.timezone = UTC/g' /etc/php.ini
$ service httpd start
$ service mysqld start
$ curl -sS https://getcomposer.org/installer | php
$ mv composer.phar /usr/bin/composer
$ rm -rf /var/www/html
$ composer create-project silverstripe/installer /var/www/html
$ wget https://raw.githubusercontent.com/BetterBrief/vagrant-skeleton/v1.0.0/_ss_environment.php -P /var/www/ --no-check-certificate
$ exit
$ exit
```
Navigate to http://localhost:8080 in browser.

Follow your steps and it works for me.

**Comment by stevie-mayhew** [comment]
@dhensby I can confirm that this works as expected too. 

Might be some custom code in there which is failing @hailwood, or perhaps your assets directory isn't writable?

**Comment by willmorgan** [comment]
Feel free to pastebin your code. I used PHP 5.3.3 for quite some time and never experienced any issues with exporting/printing gridfield data.

**Comment by hailwood** [comment]
Hey guys,

Here is the yeoman generator we use that will setup silverstripe with a vagrant box that can reproduce the issue: https://github.com/XploreNet/generator-xplorestripe

**Comment by dhensby** [comment]
@hailwood it seems to get that working you need node installed locally, amirite? I don't have node installed locally and I don't plan to at the moment.

I guess I could give this a run through with a debian box, but I'm not 100% familiar with Debian based systems and considering that this isn't widely reported, I think we have to assume it's an environment problem at the moment.

**Comment by hailwood** [comment]
Yeah you're correct there.

I guess we can put it down to environment yeah, probably something to do with debian.

**Comment by dhensby** [comment]
@hailwood if it's something to do with stock debian and stock php/apache (from their repos) then I'd agree it's a bug as we should be supporting that. I'm leaning more toward it being a problem somewhere else that is specific you your environment or codebase.

If you could package and publish/share a vagrant box that has the issue, that'd be great

**Comment by hailwood** [comment]
@dhensby https://atlas.hashicorp.com/xplore/boxes/debian-6.0.9

**Comment by dhensby** [comment]
Does that have SS installed and failing?

I meant a box packaged with the code.

**Comment by nfauchelle** [comment]
I was just thinking about this, and I wonder if it has something to do with the suhosin php module that is (I think) with Debian (I know Ubuntu used to / does have it).

That would explain why it works on CentOS.

@hailwood 
Maybe checking to see if it's enabled, disable it from checking would prove if that is what it is.

**Comment by ss23** [comment]
I'm also getting this with `PHP 5.3.3-7+squeeze25 with Suhosin-Patch (cli) (built: Feb  2 2015 16:30:12)` on Debian Squeeze-LTS.

**Comment by ss23** [comment]
The issue for me was various suhosin parameters.
Specifically, I changed:
`suhosin.request.max_varname_length = 256`

Given people will have the default options in use, I would recommend the initial fix that @hailwood suggested.

**Comment by wernerkrauss** [comment]
had a similar problem today on a debian box with php 5.3.3-7. Errorlog didn't show anything, suhosin logs to syslog!

Had to update `suhosin.get.max_name_length` to get excel export working and `suhosin.post.max_name_length` to get pagination and column sorting working.

**Comment by ARNHOE** [comment]
Suhosin was the problem for us aswell!

**Comment by ss23** [comment]
@dhensby Any chance of getting the "proper" (well, the "fuck suhosin but we'll support that crap anyway) fix merged? Looks like lots of people are hitting this.

**Comment by dhensby** [comment]
What is the fix :(

It seems like we need to stop pushing such long URL variables?

**Comment by ss23** [comment]
I was thinking `parse_str($_SERVER['QUERY_STRING'], $data);` in the short term (like in the initial post).
Or a check on install for suhosin, or not using long URL variables.

In the interest of keeping SilverStripe actually working for now, the workaround seems best to me, since this affects "stable" versions? Unless there is someone who can devote time to a proper fix.

**Comment by ARNHOE** [comment]
@dhensby Adding the following lines in php.ini fixes this.

```
suhosin.get.max_name_length = 256
suhosin.post.max_name_length = 256
```

**Comment by dhensby** [comment]
@ARNHOE we can't manipulate people's PHP.ini so this isn't a viable solution for us to implement.

@ss23 that could be a solution, but for the bootstrap we chould just re-define the GET vars... but that seems nasty...

is the problem fixed by editing https://github.com/silverstripe/silverstripe-framework/blob/3.1/core/Constants.php#L154 and adding `parse_str($_SERVER['QUERY_STRING'], $_GET);` ?

**Comment by ARNHOE** [comment]
@dhensby I understand, but any bits of information helps I guess

**Comment by dhensby** [comment]
@ARNHOE sorry, yes, good for reference for other users.

**Comment by dhensby** [comment]
I'm going to edit the title of this ticket to mention suhosin php as it'll help people searching for this issue.

**Comment by ss23** [comment]
I believe #3971 may have fixed this.

@chillu @tractorcow @dhensby Can you confirm?

**Comment by dhensby** [comment]
@tractorcow confirms this was fixed by #3971

**Comment by tractorcow** [comment]
Did I? :P It possibly did fix it.
